onnxruntime/core/graph/graph_flatbuffers_utils.cc

```diff
@@ -339,4 +340,134 @@ Status LoadAttributeOrtFormat(const fbs::Attribute& fbs_attr,
   return Status::OK();
 }
 
+#ifdef ENABLE_TRAINING_APIS
+
+Status SaveOrtValueOrtFormat(
+    const std::string& tensor_name, const OrtValue& ort_value,
+    const std::function<Status(const onnxruntime::Tensor& src_tensor, onnxruntime::Tensor& dst_tensor)> copy_tensor,
+    flatbuffers::FlatBufferBuilder& builder,
+    flatbuffers::Offset<fbs::Tensor>& fbs_tensor) {
+  // Check if the OrtValue is a tensor.
+  ORT_RETURN_IF_NOT(ort_value.IsTensor(), "Only tensor OrtValues can be saved to a checkpoint.");
+
+  const auto ort_tensor_to_fbs_tensor = [&builder](const auto& tensor_name, const auto& ort_tensor, auto& fbs_tensor) {
+    ORT_RETURN_IF(ort_tensor.IsDataTypeString(),
+                  "TensorProto_DataType_STRING is not supported while saving a tensor to ORT format.");
+
+    const auto fbs_tensor_name = builder.CreateString(tensor_name);
+    const auto fbs_tensor_doc_string = builder.CreateString(std::string());
+    const auto fbs_tensor_dims = SaveDims(builder, ort_tensor.Shape().GetDims());
+    flatbuffers::Offset<flatbuffers::Vector<uint8_t>> raw_data = builder.CreateVector(
+        static_cast<const uint8_t*>(ort_tensor.DataRaw()),
+        ort_tensor.SizeInBytes());
+
+    fbs::TensorBuilder tb(builder);
+    tb.add_name(fbs_tensor_name);
+    tb.add_doc_string(fbs_tensor_doc_string);
+    tb.add_dims(fbs_tensor_dims);
+    tb.add_data_type(static_cast<fbs::TensorDataType>(ort_tensor.GetElementType()));
+    tb.add_raw_data(raw_data);
+    fbs_tensor = tb.Finish();
+    return Status::OK();
+  };
+
+  const onnxruntime::Tensor& src_tensor = ort_value.Get<onnxruntime::Tensor>();
+
+  // Check if the tensor is on CPU. If not, we need to copy the tensor to CPU before saving it.
+  if (const auto& tensor_location = src_tensor.Location();
+      tensor_location.device.Type() != OrtDevice::CPU) {
+    InlinedVector<uint8_t> tensor_data_buffer{};
+    tensor_data_buffer.resize(src_tensor.SizeInBytes());
+    const OrtMemoryInfo cpu_alloc_info{onnxruntime::CPU, OrtDeviceAllocator};
+    onnxruntime::Tensor dst_tensor{src_tensor.DataType(), src_tensor.Shape(), tensor_data_buffer.data(), cpu_alloc_info};
+    ORT_RETURN_IF_ERROR(copy_tensor(src_tensor, dst_tensor));
+    ORT_RETURN_IF_ERROR(ort_tensor_to_fbs_tensor(tensor_name, dst_tensor, fbs_tensor));
+  } else {
+    ORT_RETURN_IF_ERROR(ort_tensor_to_fbs_tensor(tensor_name, src_tensor, fbs_tensor));
+  }
+
+  return Status::OK();
+}
+
+#define UNPACK_TESOR_WITH_TYPE(T)                                                                          \
```

Comment:
```suggestion
#define UNPACK_TENSOR_WITH_TYPE(T)                                                                          \
```

could we use MLTypeCallDispatcher instead of the switch with macros?
if we need to use this macro, can it be less context-dependent? i.e., use parameters instead of assuming certain variable names